### PR TITLE
ARROW-10928: [C++] Better Parquet error when trying to write empty struct

### PR DESCRIPTION
**Pull request by pitrou**
An empty struct type (with no child fields) is not easy to write in Parquet,
since Parquet only represents the data of leaf nodes.
We would need a way to distinguish between null and non-null (empty) struct values.
It would probably require a dummy primitive node.

Until we implement such a solution, simply raise a nice error when an empty
struct is encountered.

### CODE REVIEW

File: cpp/src/parquet/arrow/schema.cc [hunk 1]
```diff
@@ -113,12 +114,21 @@ Status StructToNode(const std::shared_ptr<::arrow::StructType>& type,
...
+    for (int i = 0; i < type->num_fields(); i++) {
+      RETURN_NOT_OK(FieldToNode(type->field(i)->name(), type->field(i), properties,
+                                arrow_properties, &children[i]));
+    }
+  } else {
+    // XXX (ARROW-10928) We could add a dummy primitive node but that would
+    // require special handling when writing and reading, to avoid column index
+    // mismatches.
+    return Status::NotImplemented(
+        "Cannot write struct type with no child fields to Parquet. "
```
Comment (emkornfield) [inline cpp/src/parquet/arrow/schema.cc:127]: would the struct field have a name here?  Consider removing "dummy" from the message.

File: cpp/src/parquet/schema.cc [hunk 2]
```diff
@@ -550,11 +550,11 @@ std::unique_ptr<Node> Unflatten(const format::SchemaElement* elements, int lengt
     int field_id = current_id++;
     const void* opaque_element = static_cast<const void*>(&element);
 
-    if (element.num_children == 0) {
+    if (!element.__isset.num_children) {
```
Comment (emkornfield) [inline cpp/src/parquet/schema.cc:553]: should this be an or?

**Review by emkornfield — CHANGES_REQUESTED** [review]
a couple of minor comments, otherwise LGTM.

File: cpp/src/parquet/arrow/schema.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/parquet/arrow/schema.cc:127]: Hmm, I can add the struct field name to the message, but I'm not sure I understand the suggestion about "dummy".

File: cpp/src/parquet/schema.cc (hunk 2 above)
Comment (pitrou) [reply cpp/src/parquet/schema.cc:553]: Hmm, why would it be an or? If `num_children == 0`, this can still be a group node.

Comment (pitrou) [reply cpp/src/parquet/schema.cc:553]: Though it seems to produce failures on some Python Parquet tests (because of legacy files?). Let me see.

**Pull request closed without merging**